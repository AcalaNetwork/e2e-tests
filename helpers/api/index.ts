import { ApiPromise } from '@polkadot/api'

export const xtokens = {
  relaychainV2: (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            id: acc,
          },
        },
      },
    },
  }),
  parachainV2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  parachainAccountId20V2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountKey20: {
              network: 'Any',
              key: acc,
            },
          },
        ],
      },
    },
  }),
  parachainV3: (paraId: number) => (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  transfer:
    (token: any, amount: any, dest: (dest: any) => any, weight: any = 'Unlimited') =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transfer(token, amount, dest(acc), weight),
  transferMulticurrencies:
    (token: any, amount: any, feeToken: any, feeAmount: any, dest: (dest: any) => any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transferMulticurrencies(
        [
          [token, amount],
          [feeToken, feeAmount],
        ],
        1,
        dest(acc),
        'Unlimited'
      ),
}

export const xcmPallet = {
  parachainV2: (parents: number, paraId: number) => ({
    V1: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  parachainV3: (parents: number, paraId: any) => ({
    V3: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  limitedReserveTransferAssetsV2:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V1: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  network: 'Any',
                  id: acc,
                },
              },
            },
          },
        },
        {
          V1: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
  limitedReserveTransferAssetsV3:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
}

export const tx = {
  xtokens,
  xcmPallet,
}

export const query = {
  balances: ({ api }: { api: ApiPromise }, address: string) => api.query.system.account(address),
  tokens:
    (token: any) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.tokens.accounts(address, token),
  assets:
    (token: number | bigint) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.assets.account(token, address),
  evm:
    (contract: string, slot: string) =>
    ({ api }: { api: ApiPromise }, _address: string) =>
      api.query.evm.accountStorages(contract, slot),
}
