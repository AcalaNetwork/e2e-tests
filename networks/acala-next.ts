import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acalaNext' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karuraNext' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala_runtime.wasm',
        kusama: './wasm/karura_runtime.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
