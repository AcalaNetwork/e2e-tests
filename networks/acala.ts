import { Config } from './types'

export type Vars = {
  relayToken: string
  relayLiquidToken: string
}

export default {
  polkadot: {
    name: 'acala' as const,
    endpoint: 'wss://acala-rpc-0.aca-api.network',
    relayToken: 'DOT',
    relayLiquidToken: 'LDOT',
  },
  kusama: {
    name: 'karura' as const,
    endpoint: 'wss://karura-rpc-0.aca-api.network',
    relayToken: 'KSM',
    relayLiquidToken: 'LKSM',
  },
  config: ({ alice, relayToken, relayLiquidToken }) => ({
    storages: {
      System: {
        account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Tokens: {
        accounts: [
          [[alice.address, { Token: relayToken }], { free: 10 * 1e12 }],
          [[alice.address, { Token: relayLiquidToken }], { free: 100 * 1e12 }],
        ],
      },
      Sudo: {
        key: alice.address,
      },
      Homa: {
        // avoid impact test outcome
        $removePrefix: ['redeemRequests', 'unbondings', 'toBondPool'],
        // so that bump era won't trigger unbond
        relayChainCurrentEra: '0x64000000',
      },
      EvmAccounts: {
        accounts: [[['0x82a258cb20e2adb4788153cd5eb5839615ece9a0'], alice.address]],
        evmAddresses: [[[alice.address], '0x82a258cb20e2adb4788153cd5eb5839615ece9a0']],
      },
      PolkadotXcm: {
        $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
      },
    },
  }),
} satisfies Config<Vars>

export const acala = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  dot: { Token: 'DOT' },
  dai: { Erc20: '0x54a37a01cd75b616d63e0ab665bffdb0143c52ae' },
  wbtc: { ForeignAsset: 5 },
} as const

export const karura = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  ksm: { Token: 'KSM' },
  usdt: { ForeignAsset: 7 },
  dai: { Erc20: '0x4bb6afb5fa2b07a5d1c499e1c3ddb5a15e709a71' },
} as const
