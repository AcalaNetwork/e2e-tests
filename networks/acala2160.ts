import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2160' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2160' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2160-dev.wasm',
        kusama: './wasm/karura-2160-dev.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
