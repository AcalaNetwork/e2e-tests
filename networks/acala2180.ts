import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2180' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2180' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2180.wasm',
        kusama: './wasm/karura-2180.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
