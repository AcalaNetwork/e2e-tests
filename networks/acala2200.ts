import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2200' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2200' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2200.wasm',
        kusama: './wasm/karura-2200.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
