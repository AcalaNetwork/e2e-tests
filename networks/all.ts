import { Config } from './types'

import acala2160Config from './acala2160'
import acalaConfig from './acala'
import hydraDXConfig from './hydraDX'
import moonbeamConfig from './moonbeam'
import polkadot9381Config from './polkadot9381'
import polkadotConfig from './polkadot'
import statemintConfig from './statemint'

const all = {
  polkadot: polkadotConfig,
  polkadot9381: polkadot9381Config,
  statemint: statemintConfig,
  acala: acalaConfig,
  acala2160: acala2160Config,
  moonbeam: moonbeamConfig,
  hydraDX: hydraDXConfig,
} satisfies Record<string, Config>

export default all
