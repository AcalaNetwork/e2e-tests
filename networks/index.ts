import { SetupOption, setupContext } from '@acala-network/chopsticks-testing'
import { connectParachains, connectVertical } from '@acala-network/chopsticks'
import dotenv from 'dotenv'

import { testingPairs } from '../helpers'

import { Config, Context, NetworkKind } from './types'
import networkDefs from './all'

dotenv.config()

const toNumber = (value: string | undefined): number | undefined => {
  if (value === undefined) {
    return undefined
  }

  return Number(value)
}

export type Network = Awaited<ReturnType<typeof setupContext>> & {
  options: SetupOption
  config: (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]
}
export type NetworkNames = (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]['name']

export const networkCreator = {} as Record<
  NetworkNames,
  (options?: Partial<SetupOption>) => (ctx: Context) => Promise<Network>
>

const relaychains = ['polkadot', 'kusama'] as const

for (const def of Object.values(networkDefs)) {
  for (const relaychain of relaychains) {
    const config = def[relaychain]
    const { endpoint, name } = config
    const upperName = name.toUpperCase()
    networkCreator[name] = (options?: Partial<SetupOption>) => async (ctx: Context) => {
      const setupConfig = (def as Config).config({
        network: relaychain,
        ...config,
        ...ctx,
      })

      const finalOptions: SetupOption = {
        timeout: 600000,
        wasmOverride: process.env[`${upperName}_WASM`],
        blockNumber: toNumber(process.env[`${upperName}_BLOCK_NUMBER`]),
        endpoint: process.env[`${upperName}_ENDPOINT`] ?? endpoint,
        db: process.env.DB_PATH,
        ...setupConfig.options,
        ...options,
      }

      const network = await setupContext(finalOptions)

      await network.dev.setStorage(setupConfig.storages)

      return {
        ...network,
        config,
        options: finalOptions,
      }
    }
  }
}

export const createContext = () => testingPairs()

export const createNetworks = async (
  networkOptions: Partial<Record<NetworkNames, Partial<SetupOption> | undefined>>,
  context = createContext()
) => {
  const ret = {} as Record<NetworkNames, Network>

  let wasmOverriden = false

  for (const [name, options] of Object.entries(networkOptions) as [NetworkNames, Partial<SetupOption> | undefined][]) {
    ret[name] = await networkCreator[name](options)(context)
    wasmOverriden ||= !!ret[name].options.wasmOverride
  }

  const relaychainName = Object.keys(ret).filter(
    (x) => x.startsWith('polkadot') || x.startsWith('kusama')
  )[0] as NetworkKind
  const { [relaychainName]: relaychain, ...parachains } = ret

  if (relaychain) {
    for (const parachain of Object.values(parachains)) {
      await connectVertical(relaychain.chain, parachain.chain)
    }
  }

  const parachainList = Object.values(parachains).map((i) => i.chain)
  if (parachainList.length > 0) {
    await connectParachains(parachainList)
  }

  if (wasmOverriden) {
    // trigger runtime upgrade if needed (due to wasm override)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
    // handle xcm version message if needed (due to wasm override triggered xcm version upgrade)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
  }

  return ret
}

// to be compatible with old code

const networks = {} as Record<NetworkNames, (options?: Partial<SetupOption>) => Promise<Network>>

for (const [name, creator] of Object.entries(networkCreator)) {
  networks[name as NetworkNames] = async (options?: Partial<SetupOption>) => creator(options)(createContext())
}

export default networks
