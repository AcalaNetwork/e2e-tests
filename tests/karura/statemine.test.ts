import { afterAll, describe, expect, it } from 'vitest'
import { connectParachains } from '@acala-network/chopsticks'

import {
  balance,
  expectJson,
  matchEvents,
  matchHrmp,
  matchSystemEvents,
  sendTransaction,
  testingPairs,
} from '../helper'
import { xTokensTransferMulticurrencies } from '../api/extrinsics'
import networks from '../networks'

describe('Karura <-> Statemine', async () => {
  const statemine = await networks.statemine()
  const karura = await networks.karura()
  await connectParachains([statemine.chain, karura.chain])

  const { alice, bob } = testingPairs()

  afterAll(async () => {
    await statemine.teardown()
    await karura.teardown()
  })

  it('0. Statemine transfer assets to Karura', async () => {
    // give Alice some KSM and USDt
    await statemine.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 1000 * 1e10 } }]],
      },
      Assets: {
        Account: [[[1984, alice.address], { balance: 1000e6 }]],
      },
    })

    // ensure balance was given
    expect(await balance(statemine.api, alice.address)).toMatchInlineSnapshot(`
      {
        "feeFrozen": 0,
        "free": 10000000000000,
        "miscFrozen": 0,
        "reserved": 0,
      }
    `)
    expect((await statemine.api.query.assets.account(1984, alice.address)).toHuman()).toMatchInlineSnapshot(`
      {
        "balance": "1,000,000,000",
        "extra": null,
        "isFrozen": false,
        "reason": "Consumer",
      }
    `)

    await statemine.api.tx.polkadotXcm
      .limitedReserveTransferAssets(
        {
          V0: {
            X2: [{ Parent: null }, { Parachain: 2000 }],
          },
        },
        {
          V0: {
            X1: {
              AccountId32: {
                network: 'Any',
                id: alice.addressRaw,
              },
            },
          },
        },
        {
          V0: [
            {
              ConcreteFungible: {
                amount: 10e6,
                id: {
                  X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }],
                },
              },
            },
          ],
        },
        0,
        { Unlimited: null }
      )
      .signAndSend(alice)

    await statemine.chain.newBlock()
    await karura.chain.newBlock()

    expect(await balance(statemine.api, alice.address)).toMatchInlineSnapshot(`
      {
        "feeFrozen": 0,
        "free": 9999937253833,
        "miscFrozen": 0,
        "reserved": 0,
      }
    `)
    await matchSystemEvents(statemine, {
      section: 'polkadotXcm',
      method: 'Attempted',
    })

    await matchSystemEvents(karura, {
      section: 'xcmpQueue',
      method: 'Success',
    })

    // ensure Alice got the money
    expect((await karura.api.query.tokens.accounts(alice.address, { ForeignAsset: '7' })).toHuman())
      .toMatchInlineSnapshot(`
      {
        "free": "9,999,192",
        "frozen": "0",
        "reserved": "0",
      }
    `)
  })

  it('1. Karura transfer assets to Statemine', async () => {
    await karura.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 1000 * 1e10 } }]],
      },
      Tokens: {
        Accounts: [[[alice.address, { Token: 'KSM' }], { free: 10 * 1e12 }]],
      },
    })
    expect(await balance(statemine.api, alice.address)).toMatchInlineSnapshot(`
      {
        "feeFrozen": 0,
        "free": 9999937253833,
        "miscFrozen": 0,
        "reserved": 0,
      }
    `)
    expect(await balance(karura.api, alice.address)).toMatchInlineSnapshot(`
      {
        "feeFrozen": 0,
        "free": 10000000000000,
        "miscFrozen": 0,
        "reserved": 0,
      }
    `)
    expect((await karura.api.query.tokens.accounts(alice.address, { ForeignAsset: '7' })).toHuman())
      .toMatchInlineSnapshot(`
      {
        "free": "9,999,192",
        "frozen": "0",
        "reserved": "0",
      }
    `)
    const tx = await sendTransaction(
      xTokensTransferMulticurrencies(karura.api, '7', '9999192', '1000', bob.addressRaw).signAsync(alice)
    )
    await karura.chain.newBlock()
    await statemine.chain.newBlock()

    expectJson(await karura.api.query.tokens.accounts(alice.address, { ForeignAsset: '7' })).toMatchInlineSnapshot(`
      {
        "free": 0,
        "frozen": 0,
        "reserved": 0,
      }
    `)

    await matchEvents(tx.events, 'xTokens', 'TransferredMultiAssets', 'xcmpQueue')
    await matchHrmp(karura)
    await matchSystemEvents(statemine, {
      section: 'xcmpQueue',
      method: 'Success',
    })
    expectJson(await statemine.api.query.assets.account(1984, bob.address)).toMatchInlineSnapshot(`
      {
        "balance": 9999192,
        "extra": null,
        "isFrozen": false,
        "reason": {
          "sufficient": null,
        },
      }
    `)
    // expectEvent(await statemine.api.query.system.events(), {
    //   event: expect.objectContaining({
    //     section: 'xcmpQueue',
    //     method: 'Success'
    //   })
    // })
  })
})
