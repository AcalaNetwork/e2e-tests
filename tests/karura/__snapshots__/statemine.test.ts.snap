// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Karura <-> Statemine > 0. Statemine transfer assets to Karura > system events 1`] = `
[
  {
    "data": [
      {
        "Complete": "(redacted)",
      },
    ],
    "method": "Attempted",
    "section": "polkadotXcm",
  },
]
`;

exports[`Karura <-> Statemine > 0. Statemine transfer assets to Karura > system events 2`] = `
[
  {
    "data": {
      "messageHash": "(redacted)",
      "weight": {
        "proofSize": "(redacted)",
        "refTime": "(redacted)",
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`Karura <-> Statemine > 1. Karura transfer assets to Statemine > events 1`] = `
[
  {
    "data": {
      "messageHash": "(redacted)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": "(redacted)",
          },
          "id": {
            "Concrete": {
              "interior": "Here",
              "parents": "(redacted)",
            },
          },
        },
        {
          "fun": {
            "Fungible": "(redacted)",
          },
          "id": {
            "Concrete": {
              "interior": {
                "X3": [
                  {
                    "Parachain": "(redacted)",
                  },
                  {
                    "PalletInstance": "(redacted)",
                  },
                  {
                    "GeneralIndex": "(redacted)",
                  },
                ],
              },
              "parents": "(redacted)",
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": "(redacted)",
            },
            {
              "AccountId32": {
                "id": "(redacted)",
                "network": "Any",
              },
            },
          ],
        },
        "parents": "(redacted)",
      },
      "fee": {
        "fun": {
          "Fungible": "(redacted)",
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": "(redacted)",
          },
        },
      },
      "sender": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
]
`;

exports[`Karura <-> Statemine > 1. Karura transfer assets to Statemine > hrmp 1`] = `
[
  {
    "data": "0x0002100008000002043205011f00624d620200010000028c86470a1300010000028c8647000d01000800010100d17c2d7823ebf260fd138f2d7e27d114c0145d968b5ff5006125f2414fadae69",
    "recipient": 1000,
  },
]
`;

exports[`Karura <-> Statemine > 1. Karura transfer assets to Statemine > system events 1`] = `
[
  {
    "data": {
      "messageHash": "(redacted)",
      "weight": {
        "proofSize": "(redacted)",
        "refTime": "(redacted)",
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;
